src/util.cpp

```diff
@@ -373,27 +373,50 @@ int LogPrintStr(const std::string &str)
     return ret;
 }
 
+/** A map that contains all the currently held directory locks. After
+ * successful locking, these will be held here until the global destructor
+ * cleans them up and thus automatically unlocks them, or ReleaseDirectoryLocks
+ * is called.
+ */
+static std::map<std::string, std::unique_ptr<boost::interprocess::file_lock>> dir_locks;
+/** Mutex to protect dir_locks. */
+static std::mutex cs_dir_locks;
+
 bool LockDirectory(const fs::path& directory, const std::string lockfile_name, bool probe_only)
 {
+    std::lock_guard<std::mutex> ulock(cs_dir_locks);
     fs::path pathLockFile = directory / lockfile_name;
-    FILE* file = fsbridge::fopen(pathLockFile, "a"); // empty lock file; created if it doesn't exist.
+
+    // If a lock for this directory already exists in the map, don't try to re-lock it
+    if (dir_locks.count(pathLockFile.string())) {
+        return true;
+    }
+
+    // Create empty lock file if it doesn't exist.
+    FILE* file = fsbridge::fopen(pathLockFile, "a");
     if (file) fclose(file);
 
     try {
-        static std::map<std::string, boost::interprocess::file_lock> locks;
-        boost::interprocess::file_lock& lock = locks.emplace(pathLockFile.string(), pathLockFile.string().c_str()).first->second;
-        if (!lock.try_lock()) {
+        auto lock = MakeUnique<boost::interprocess::file_lock>(pathLockFile.string().c_str());
+        if (!lock->try_lock()) {
```

Comment:
In the probe_only case, this won't be unlocked when ```lock``` destructs, as interprocess::file_lock isn't RAII. Was it intended to work that way?

Reply:
Eh are you sure? the whole intent of keeping the things in a std::map is that they get unlocked when they are destructed. Anecdotally this seems to work on windows, at least, the call to `ReleaseDirectoryLocks()` seems to release the locks. I can try further.

Reply:
That the documentation is confusing here:

> A file locking is a class that has process lifetime. This means that if a process holding a file lock ends or crashes, the operating system will automatically unlock it. This feature is very useful in some situations where we want to assure automatic unlocking even when the process crashes and avoid leaving blocked resources in the system. A file lock is constructed using the name of the file as an argument:

This would suggest we don't even need to hold on to the `file_lock` after locking them, as the locks always have process lifetime. But the test outcomes seem to suggest RAII behavior.

Reply:
Grr, the "theuni requested changes" is a bit overstated... There are a few fixes here and I can't produce any real-world issues, so utACK from me for 0.16 for the sake of not dragging it out.

----

Further discussion for master, or 0.16 if you still feel like messing with it:


Strange, your tests results (RAII behavior) don't line up with what I see from the docs/code. I'll play with tests.

You mentioned:
> The `FILE* file = fsbridge::fopen(pathLockFile, "a");`  wipes the 'we own this lock' administration

And from the Boost docs:
> In POSIX, when two file descriptors are used to lock a file if a descriptor is closed, all file locks set by the calling process are cleared. 

Is it possible that it was the fclose rather than the fopen? Because that would explain everything neatly, I think. In the ```probe_only``` case, You create a lock, lock it, then destroy it. Then on the next invocation you fopen/fclose it, which basically gives a clean start before you create a new lock, lock again, etc.

To be clear, my concern is that in the near future we might adapt LockDirectory() and add UnlockDirectory() for some feature like multi-wallet-dirs, only to be hit with a hard-to-diagnose bug when UnlockDirectory() quietly doesnt' work as intended.

The [boost docs](http://www.boost.org/doc/libs/1_66_0/doc/html/interprocess/synchronization_mechanisms.html#interprocess.synchronization_mechanisms.file_lock.file_lock_scoped_locks) suggest using RAII generics here:
> scoped_lock and sharable_lock can be used to make file locking easier in the presence of exceptions, just like with mutexes.

And file_lock's destructor is pretty straightforward, I don't see how it could be doing any unlocking:
```c++
// m_file_hnd: typedef int (void* for win)
// ipcdetail::close_file: ::close (CloseHandle() for win)
inline file_lock::~file_lock()
{
   if(m_file_hnd != ipcdetail::invalid_file()){
      ipcdetail::close_file(m_file_hnd);
      m_file_hnd = ipcdetail::invalid_file();
   }
}
```

Reply:
> Is it possible that it was the fclose rather than the fopen? Because that would explain everything neatly, I think. In the probe_only case, You create a lock, lock it, then destroy it. Then on the next invocation you fopen/fclose it, which basically gives a clean start before you create a new lock, lock again, etc.

Whoa, sneaky. Yes, that's possible. Let's try the tests that by using an already-existing directory with an existing lock file, and remove that part.